lib/utils/github/handle-updates.ts

```diff
@@ -0,0 +1,137 @@
+import {Webhooks} from '@octokit/webhooks'
+import {GITHUB} from '~/constants'
+import {Repository} from '~/types'
+import Weaviate from '../embeddings/db'
+import {getMainBranch} from '../github'
+
+/**
+ * Handles if a repository(s) is added or removed for an existing installation
+ */
+export default function handleUpdates(webhook: Webhooks<unknown>) {
+	webhook.on(
+		['installation_repositories.added', 'installation_repositories.removed'],
+		async ({payload}) => {
+			// Added or removed repos
+			const {
+				repositories_added: addedRepos,
+				repositories_removed: removedRepos,
+				installation: {
+					account: {
+						id: githubOrgId,
+						login: githubOrgSlug,
+						avatar_url: githubOrgImage,
+						type
+					}
+				},
+				sender: {id: githubUserId, login: userName}
+			} = payload
+
+			// Get user
+			const user = await prisma.user.findUnique({
+				where: {
+					githubUserId: githubUserId.toString()
+				},
+				select: {
+					id: true,
+					addProject: {
+						take: 1,
+						orderBy: {createdAt: 'desc'}
+					}
+				}
+			})
+
+			if (!user?.id)
+				return new Response(`Could not find user ${userName}`, {
+					status: 500
+				})
+
+			try {
+				let createProjectsAndOrg = null
+
+				// If repo(s) is under an organization (not personal), create projects with corresponding organization
+				if (type === 'Organization')
+					createProjectsAndOrg = prisma.organization.upsert({
+						where: {githubOrganizationId: githubOrgId.toString()},
+						create: {
+							githubOrganizationId: githubOrgId.toString(),
+							slug: githubOrgSlug,
+							name: githubOrgSlug,
+							image: githubOrgImage,
+							projects: {
+								createMany: {
+									data: addedRepos.map((repo: Repository) => ({
+										githubProjectId: repo.id.toString(),
+										slug: repo.name,
+										name: repo.name,
+										private: repo.private,
+										createdBy: user.id,
+										teamId: user.addProject[0].teamId
+									})),
+									skipDuplicates: true
+								}
+							}
+						},
+						update: {
+							slug: githubOrgSlug,
+							name: githubOrgSlug,
+							image: githubOrgImage,
+							projects: {
+								createMany: {
+									data: addedRepos.map((repo: Repository) => ({
+										githubProjectId: repo.id.toString(),
+										slug: repo.name,
+										name: repo.name,
+										private: repo.private,
+										createdBy: user.id,
+										teamId: user.addProject[0].teamId
+									})),
+									skipDuplicates: true
+								}
+							}
+						}
+					})
+				// If repo(s) is for an individual user, just create the projects directly without an organization
+				else if (type === 'User')
```

Comment:
The user vs org changes seem to introduce a lot of complexity. Want to decouple this into 2+ PRs?
1. Type-safe GitHub webhooks
2. Org support

If that's substantially more work, it might not be tenable. Otherwise, I wonder if we can keep this PR manageable in size.

Reply:
Unfortunately they are heavily coupled, most of the work is in one file `api/webhook/github/route.ts` so it's hard to isolate the two.